ext/mbstring/mbstring.c

```diff
@@ -3860,31 +3860,34 @@ PHP_FUNCTION(mb_get_info)
 }
 /* }}} */
 
-static int mbfl_filt_check_errors(int c, void* data)
-{
-	if (c == MBFL_BAD_INPUT) {
-		(*((mbfl_convert_filter**)data))->num_illegalchar++;
-	}
-	return 0;
-}
-
 MBSTRING_API int php_mb_check_encoding(const char *input, size_t length, const mbfl_encoding *encoding)
 {
-	mbfl_convert_filter *filter = mbfl_convert_filter_new(encoding, &mbfl_encoding_wchar, mbfl_filt_check_errors, NULL, &filter);
-
-	while (length--) {
-		unsigned char c = *input++;
-		(filter->filter_function)(c, filter);
-		if (filter->num_illegalchar) {
-			mbfl_convert_filter_delete(filter);
+	uint32_t wchar_buf[128];
+	unsigned char *in = (unsigned char*)input;
+	unsigned int state = 0;
+
+	/* If the input string is not encoded in the given encoding, there is a significant chance
+	 * that this will be seen in the first bytes. Therefore, rather than converting an entire
+	 * buffer of 128 codepoints, convert and check just a few codepoints first */
```

Comment:
I think the prior for mb_check_encoding is that the string is valid, so we probably don't need to optimize for this case.

Reply:
Hmm. You don't feel it's worth a few extra lines for a major performance boost in that case?

If not, I can remove them.

Reply:
Don't care strongly about this, just skeptical that it's relevant for real-world usage -- it probably only makes benchmarks look nicer ;)

Reply:
I like pretty benchmarks.